sdk/include/xdk/extypes.h

```diff
@@ -346,3 +346,14 @@ typedef struct _ZONE_HEADER {
 #define EX_PUSH_LOCK ULONG_PTR
 #define PEX_PUSH_LOCK PULONG_PTR
 $endif (_NTIFS_)
+
+$if (_WINNT_ || _WDMDDK_)
+#if (NTDDI_VERSION >= NTDDI_VISTA)
```

Comment:
Isn't this introduced only on Windows 8+?

Reply:
```C
#if (NTDDI_VERSION >= NTDDI_LONGHORN)
typedef struct _SYSTEM_BOOT_ENVIRONMENT_INFORMATION
{
    GUID BootIdentifier;
    FIRMWARE_TYPE FirmwareType;
#if (NTDDI_VERSION >= NTDDI_WIN8)
    ULONGLONG BootFlags;
#endif
} SYSTEM_BOOT_ENVIRONMENT_INFORMATION, *PSYSTEM_BOOT_ENVIRONMENT_INFORMATION;
#endif
```

`FIRMWARE_TYPE` is available since NT 6.0 in KM (wdm.h), and NT 6.2 in UM (winnt.h), so how about:
```C
$if (_WDMDDK_)
#if (NTDDI_VERSION >= NTDDI_VISTA)
$endif (_WDMDDK_)
$if (_WINNT_)
#if (NTDDI_VERSION >= NTDDI_WIN8)
$endif (_WINNT_)
$if (_WINNT_ || _WDMDDK_)
typedef enum _FIRMWARE_TYPE {
    FirmwareTypeUnknown,
    FirmwareTypeBios,
    FirmwareTypeUefi,
    FirmwareTypeMax
} FIRMWARE_TYPE, *PFIRMWARE_TYPE;
#endif /* (NTDDI_VERSION >= NTDDI_VISTA) or (NTDDI_VERSION >= NTDDI_WIN8) */
$endif (_WINNT_ || _WDMDDK_)
```
Or no `#if` guarded like Windows SDK/WDK?

Reply:
> FIRMWARE_TYPE is available since NT 6.0 in KM (wdm.h) [...]

ok I see. Because I actually checked for the existence of GetFirmwareType function instead in kernel32/kernelbase, and it wasn't there on windows 7.

Reply:
Some NT6.0 build (xxx_vista) have `SYSTEM_BOOT_ENVIRONMENT_INFORMATION`, they also need `FIRMWARE_TYPE`. They should be guarded by `(NTDDI_VERSION >= NTDDI_VISTA)` or no guard together.

Reply:
ok let's just keep it simple as it currently is, with the ntddi_version >= vista.